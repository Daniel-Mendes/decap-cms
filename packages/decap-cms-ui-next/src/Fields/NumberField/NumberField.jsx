import React, { useRef, useState } from 'react';
import styled from '@emotion/styled';
import { css } from '@emotion/react';

import Field from '../../Field';
import { IconButton } from '../../Buttons';
import { StyledInput as StyledTextInput } from '../TextField';

const StyledNumberInput = styled(StyledTextInput)`
  padding-right: 3rem;

  -webkit-appearance: textfield;
  -moz-appearance: textfield;
  appearance: textfield;

  &::-webkit-inner-spin-button,
  &::-webkit-outer-spin-button {
    -webkit-appearance: none;
    margin: 0;
  }
`;

const NumberInputStepper = styled.div`
  display: flex;
  flex-direction: column;

  position: absolute;
  right: 0.5rem;
  top: 1.85rem;
`;

const NumberIncrementStepper = styled(IconButton)`
  rotate: 180deg;

  ${({ size }) => css`
    width: ${size === 'xs' ? 1.25 : 2}rem;
    height: ${size === 'xs' ? 1 : 2}rem;
  `};
`;

const NumberDecrementStepper = styled(IconButton)`
  ${({ size }) => css`
    width: ${size === 'xs' ? 1.25 : 2}rem;
    height: ${size === 'xs' ? 1 : 2}rem;
  `};
`;

function NumberField({
  name,
  label,
  status,
  placeholder,
  description,
  step,
  min,
  max,
  onChange,
  children,
  value,
  focus,
  inline,
  error,
  errors,
  ...props
}) {
  const ref = useRef(null);
  const [inputFocus, setInputFocus] = useState();

  return (
    <Field
      name={name}
      label={label}
      status={status}
      placeholder={placeholder}
      description={description}
      focus={focus || inputFocus}
      inline={inline}
      error={error}
      errors={errors}
      icon={
        <NumberInputStepper>
          <NumberIncrementStepper
            icon="chevron-down"
            size="xs"
            onClick={() => ref.current.stepUp()}
          />
          <NumberDecrementStepper
            icon="chevron-down"
            size="xs"
            onClick={() => ref.current.stepDown()}
          />
        </NumberInputStepper>
      }
      {...props}
    >
      <StyledNumberInput
        name={name}
        type="number"
        value={value}
        step={step}
        min={min}
        max={max}
        ref={ref}
        onChange={e => onChange(e)}
        onFocus={() => setInputFocus(true)}
        onBlur={() => setInputFocus(false)}
      />
      {children && children}
    </Field>
  );
}

export default NumberField;
