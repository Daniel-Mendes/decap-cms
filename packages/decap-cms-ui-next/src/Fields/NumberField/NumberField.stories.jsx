import React from 'react';
import styled from '@emotion/styled';
import { useArgs } from '@storybook/preview-api';

import NumberField from '.';

export default {
  title: 'Fields/NumberField',
  component: NumberField,
  argTypes: {
    errors: {
      control: 'object',
      if: {
        arg: 'error',
      },
    },
  },
  args: {
    label: 'Number Field Label',
    placeholder: 'Type something here',
    description: 'This is a description',
    status: 'Required',
    value: 10,
    step: 1,
    min: 0,
    max: 100,
    inline: false,
    error: false,
    errors: [
      {
        message: 'Error message.',
      },
    ],
  },
};

const StyledNumberField = styled(NumberField)`
  width: 33vw;
`;

export function _NumberField(args) {
  const [{ value }, updateArgs] = useArgs();

  function handleChange(e) {
    updateArgs({ value: e.target.value });
  }

  return <StyledNumberField {...args} value={value} onChange={handleChange} />;
}
