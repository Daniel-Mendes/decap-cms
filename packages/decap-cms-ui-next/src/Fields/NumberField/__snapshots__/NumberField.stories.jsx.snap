// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fields/NumberField _NumberField smoke-test 1`] = `"<div class="e96xuqg0 css-ia7mgt-FieldWrap-StyledNumberField e1x0ipyd6"><div class="css-14uaz9q-FieldInside-FieldInside e1x0ipyd5"><label class="css-r9gw8d-Label-StyledLabel e1x0ipyd3">Number Field Label <div color="neutral" radius="large" class="css-6930ng-BadgeWrap-BadgeWrap epfx4ki3">Required</div></label><div class="css-dj548v-ChildrenWrap e1x0ipyd4"><input type="number" step="1" min="0" max="100" class="css-u2yb3u-StyledInput-StyledNumberInput ek12dbe3" value="10"></div><div class="css-11x906l-NumberInputStepper ek12dbe2"><button class="ek12dbe1 css-1l49kfj-IconButtonWrap-NumberIncrementStepper-NumberIncrementStepper e1s566520"><svg xmlns="http://www.w3.org/2000/svg" width="12" height="12" class="" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"><polyline points="6 9 12 15 18 9"></polyline></svg></button><button class="ek12dbe0 css-vie5fw-IconButtonWrap-NumberDecrementStepper-NumberDecrementStepper e1s566520"><svg xmlns="http://www.w3.org/2000/svg" width="12" height="12" class="" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"><polyline points="6 9 12 15 18 9"></polyline></svg></button></div><p class="css-nalsj0-StyledDescription e1x0ipyd1">This is a description</p></div></div>"`;
